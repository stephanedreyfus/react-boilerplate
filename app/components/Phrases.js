import React from 'react';
import styled from 'styled-components';

const mainColor = 'mediumvioletred';

const Wrapper = styled.section`
  background: #e0e2e5;
  height: 20em;
  width: 42%;
  border-radius: 6px;
  border: 0.1em solid ${mainColor};
  box-shadow: 5px 6px 8px ${mainColor};
  margin: 3em auto auto auto;
`;

const DisplayTitle = styled.div`
  font-family: 'Montserrat', sans-serif;
  color: ${mainColor};
  text-align: center;
  border-bottom: 0.1em solid ${mainColor};
  margin: 1em 3em;
`;

const Display = styled.div`
  font-family: 'Roboto', sans-serif;
  width: 85%;
  padding: 1em;
  margin: 0.5em 1em;
  border-radius: 10px;
  border-width: 1px;
`;

// Bring in array of phrases. Map over and display after component did mount.
// Display "No phrases yet." if no phrases.
// Display "Loading..." while phrases are sought.

export default function Phrases() {
  return (
    <Wrapper>
      <DisplayTitle>Previously Entered Phrases</DisplayTitle>
      <Display>Space holder.</Display>
    </Wrapper>
  );
}
