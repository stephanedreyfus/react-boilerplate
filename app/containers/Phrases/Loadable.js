/**
 * Asynchronously loads the component for Phrases
 */

import loadable from 'utils/loadable';

export default loadable(() => import('./index'));
