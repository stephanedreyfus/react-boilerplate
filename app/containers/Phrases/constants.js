/*
 * Phrase constants for adding phrases to the database, and
 * viewing all phrases in database.
 */

export const ADD_PHRASE = 'ADD_PHRASE';
export const GET_PHRASES = 'GET_PHRASES';
