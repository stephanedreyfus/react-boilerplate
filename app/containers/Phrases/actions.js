/*
 * Phrase actions
 */

import { ADD_PHRASE } from './constants';

export function addPhrase(phrase) {
  return {
    type: ADD_PHRASE,
    payload: phrase,
  };
}
