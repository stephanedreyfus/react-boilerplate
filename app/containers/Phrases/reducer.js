/*
 * Phrase reducer
 */
import produce from 'immer';
import { ADD_PHRASE } from './constants';

export const initialState = [];

// FIXME Look at immer docs for state change documentation.
/* eslint-disable default-case, no-param-reassign */
const phraseReducer = (state = initialState, action = {}) =>
  produce(state, draft => {
    console.log('Made it into reducer');
    switch (action.type) {
      case ADD_PHRASE:
        draft.newPhrase = action.payload;
        break;
      default:
        return state;
    }
  });

export default phraseReducer;
