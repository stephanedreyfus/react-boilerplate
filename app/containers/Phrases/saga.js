import { takeLatest, put } from 'redux-saga/effects';
import axios from 'axios';

import { addPhraseError, addPhraseSuccess } from 'containers/Phrases/actions';
import { ADD_PHRASE } from './constants';
import { API_URL } from '../../utils/constants';

function* postPhrase(data) {
  try {
    const res = yield axios.post(`${API_URL}phrases`, {
      phrase: data.payload,
    });
    yield put(addPhraseSuccess(res.data));
  } catch (err) {
    yield put(addPhraseError(err));
  }
}

export default function* postPhraseWatcher() {
  yield takeLatest(ADD_PHRASE, postPhrase);
}
