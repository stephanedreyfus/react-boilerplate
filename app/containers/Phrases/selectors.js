import { createSelector } from 'reselect';
import { initialState } from './reducer';

/**
 * Direct selector to the string state domain
 */

const selectStringDomain = state => state.string || initialState;

/**
 * Other specific selectors
 */

/**
 * Default selector used by String
 */

const makeSelectString = () =>
  createSelector(
    selectStringDomain,
    substate => substate,
  );

export default makeSelectString;
export { selectStringDomain };
