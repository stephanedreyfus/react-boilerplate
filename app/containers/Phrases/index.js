/**
 * Container for adding phrases.
 * Also displays most recently added phrase.
 */

import React, { memo } from 'react';
import { createStructuredSelector } from 'reselect';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { useInjectReducer } from 'utils/injectReducer';
import { useInjectSaga } from 'utils/injectSaga';
import saga from './saga';
import reducer from './reducer';

import {
  makeSelectPhrase,
  makeSelectLoading,
  makeSelectError,
} from './selectors';
import { addPhrase } from './actions';

import DisplayField from '../../components/DisplayField';
import InputForm from '../../components/InputForm';
import {
  Wrapper,
  DisplayTitle,
  NoPhrase,
} from '../../components/Styling/PhrasesStyle';

const key = 'phrases';
let loaded;
let notLoaded;

export function Phrases({ phrase, loading, sendAddPhrase }) {
  useInjectReducer({ key, reducer });
  useInjectSaga({ key, saga });

  if (loading) {
    // FUTUREFIX Replace Loading phrase with loading graphic component
    // If no phrases are loaded, render the following:
    notLoaded = <NoPhrase>No Phrase Yet</NoPhrase>;
  } else {
    // If phrases have been loaded, render the following:
    loaded = (
      <>
        <DisplayField phrases={phrase} />
      </>
    );
  }

  return (
    <Wrapper>
      <DisplayTitle>Craft a Phrase</DisplayTitle>
      <InputForm addPhrase={sendAddPhrase} />
      <DisplayTitle>Most Recently Added</DisplayTitle>
      {loading === true ? notLoaded : loaded}
    </Wrapper>
  );
}

Phrases.propTypes = {
  loading: PropTypes.bool,
  phrase: PropTypes.array,
  sendAddPhrase: PropTypes.func,
};

const mapStateToProps = createStructuredSelector({
  phrase: makeSelectPhrase(),
  loading: makeSelectLoading(),
  error: makeSelectError(),
});

const mapDispatchToProps = {
  sendAddPhrase: phrase => addPhrase(phrase),
};

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export default compose(
  withConnect,
  memo,
)(Phrases);
