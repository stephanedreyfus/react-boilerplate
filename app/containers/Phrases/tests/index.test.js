import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { browserHistory } from 'react-router-dom';

import Phrases from '../index';
import configureStore from '../../../configureStore';

describe('<Phrases />', () => {
  let store;

  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  it.skip('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    const dispatch = jest.fn();
    render(<Phrases dispatch={dispatch} />);
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <Phrases />
        </IntlProvider>
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });
});
