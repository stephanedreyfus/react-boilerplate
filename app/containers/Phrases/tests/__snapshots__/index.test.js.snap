// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Phrases /> Should render and match the snapshot 1`] = `
<section
  class="PhrasesStyle__Wrapper-sc-13tp1yz-0 cmakpc"
>
  <div
    class="PhrasesStyle__DisplayTitle-sc-13tp1yz-1 gxwBMv"
  >
    Craft a Phrase
  </div>
  <form
    class="InputStyle__Form-sc-1fndb37-0 jGAOnu"
  >
    <div
      class="InputStyle__LabelInputWrapper-sc-1fndb37-1 dkuNvw"
    >
      <label
        class="InputStyle__Label-sc-1fndb37-2 YEXxD"
        for="phrase"
      >
        Enter a Phrase Below
      </label>
      <input
        class="InputStyle__Input-sc-1fndb37-3 dGvyjB"
        id="phrase"
        placeholder="Write anything at all!"
        type="text"
        value=""
      />
    </div>
    <button
      class="InputStyle__PhraseButton-sc-1fndb37-4 cEzmui"
    >
      Add Phrase!
    </button>
  </form>
  <div
    class="PhrasesStyle__DisplayTitle-sc-13tp1yz-1 gxwBMv"
  >
    Most Recently Added
  </div>
  <h4
    class="PhrasesStyle__NoPhrase-sc-13tp1yz-5 ixKFje"
  >
    No Phrase Yet
  </h4>
</section>
`;
